kubric/core.py

```diff
@@ -0,0 +1,212 @@
+  scale = kt.Vector3D(default_value=(1., 1., 1.))
+
+  up = tl.CaselessStrEnum(['X', 'Y', 'Z', '-X', '-Y', '-Z'], default_value='Y')
+  front = tl.CaselessStrEnum(['X', 'Y', 'Z', '-X', '-Y', '-Z'], default_value='-Z')
+
+  def look_at(self, target):
+    direction = mathutils.Vector(target) - mathutils.Vector(self.position)
+    self.quaternion = direction.to_track_quat(self.front.upper(), self.up.upper())
+
+
+class PhysicalObject(Object3D):
+  velocity = kt.Vector3D(default_value=(0., 0., 0.))
+  angular_velocity = kt.Vector3D(default_value=(0., 0., 0.))
+
+  static = tl.Bool(False)
+  mass = tl.Float(1.0)
+  friction = tl.Float(0.0)
+  restitution = tl.Float(0.5)
+
+  bounds = tl.Tuple(kt.Vector3D(), kt.Vector3D(),
+                    default_value=((0., 0., 0.), (0., 0, 0.)))
+
+  @validate('mass')
+  def _valid_mass(self, proposal):
+    mass = proposal['value']
+    if mass < 0:
+      raise tl.TraitError(f'mass cannot be negative ({mass})')
+    return mass
+
+  @validate('friction')
+  def _valid_friction(self, proposal):
+    friction = proposal['value']
+    if friction < 0:
+      raise tl.TraitError(f'friction cannot be negative ({friction})')
+    return friction
+
+  @validate('friction')
+  def _valid_friction(self, proposal):
+    friction = proposal['value']
+    if friction < 0:
+      raise tl.TraitError(f'friction cannot be negative ({friction})')
+    return friction
+
+  @validate('bounds')
+  def _valid_bounds(self, proposal):
+    lower, upper = proposal['value']
+    for l, u in zip(lower, upper):
+      if l > u:
+        raise tl.TraitError(f'lower bound cannot be larger than upper bound ({lower} !<= {upper})')
+    return lower, upper
+
+
+class FileBasedObject(PhysicalObject):
+  asset_id = tl.Unicode()
+
+  simulation_filename = tl.Unicode()   # TODO: use pathlib.Path instead
+  render_filename = tl.Unicode()       # TODO: use pathlib.Path instead
+
+  material = tl.Instance(Material, allow_none=True, default_value=None)
+
+
+# ## ### ####  Lights  #### ### ## #
+
+class Light(Object3D):
+  color = kt.RGB(default_value=Color.from_name('white').rgb)
+  intensity = tl.Float(1.)
+
+
+class DirectionalLight(Light):
+  shadow_softness = tl.Float(0.2)
+
+
+class RectAreaLight(Light):
+  width = tl.Float(1)
+  height = tl.Float(1)
+
+
+class PointLight(Light):
+  pass
+
+
+# ## ### ####  Cameras  #### ### ## #
+
+class Camera(Object3D):
+  pass
+
+
+class PerspectiveCamera(Camera):
+  focal_length = tl.Float(50)
+  sensor_width = tl.Float(36)
+
+
+class OrthographicCamera(Camera):
+  pass
+
+
+# ## ### ####  Scene  #### ### ## #
+
+class Scene(Asset):
+  frame_start = tl.Integer(0)
+  frame_end = tl.Integer(48)
+
+  frame_rate = tl.Integer(24)
+  step_rate = tl.Integer(240)
+
+  camera = tl.Instance(Camera, allow_none=True, default_value=None)
+  resolution = tl.Tuple(tl.Integer(), tl.Integer(), default_value=(512, 512))
+
+  gravity = kt.Vector3D(default_value=(0, 0, -10.))
+
+
+class AttributeSetter:
```

Comment:
Should this be here or in `blender.py` until it's used by multiple backends?

Reply:
We could move it there, though I am not sure in what way that would be better. 
I intended to also use it for the simulator, but I ended up having to write specialised setters because pybullet.

Reply:
let's revisit this once we discuss the setters in detail